ssg/templates.py

```diff
@@ -30,349 +29,32 @@
-def kernel_module_disabled(data, lang):
-    return data
-
-
-@template(["anaconda", "oval"])
-def mount(data, lang):
-    data["pointid"] = ssg.utils.escape_id(data["mountpoint"])
-    return data
-
-
-def _mount_option(data, lang):
-    if lang == "oval":
-        data["pointid"] = ssg.utils.escape_id(data["mountpoint"])
-    else:
-        data["mountoption"] = re.sub(" ", ",", data["mountoption"])
-    return data
-
-
-@template(["anaconda", "ansible", "bash", "oval"])
-def mount_option(data, lang):
-    return _mount_option(data, lang)
-
-
-@template(["ansible", "bash", "oval"])
-def mount_option_remote_filesystems(data, lang):
-    if lang == "oval":
-        data["mountoptionid"] = ssg.utils.escape_id(data["mountoption"])
-    return _mount_option(data, lang)
-
-
-@template(["anaconda", "ansible", "bash", "oval"])
-def mount_option_removable_partitions(data, lang):
-    return data
-
-
-@template(["anaconda", "ansible", "bash", "oval", "puppet"])
-def package_installed(data, lang):
-    if "evr" in data:
-        evr = data["evr"]
-        if evr and not re.match(r'\d:\d[\d\w+.]*-\d[\d\w+.]*', evr, 0):
-            raise RuntimeError(
-                "ERROR: input violation: evr key should be in "
-                "epoch:version-release format, but package {0} has set "
-                "evr to {1}".format(data["pkgname"], evr))
-    return data
-
-
-@template(["ansible", "bash", "oval"])
-def sysctl(data, lang):
-    data["sysctlid"] = ssg.utils.escape_id(data["sysctlvar"])
-    if not data.get("sysctlval"):
-        data["sysctlval"] = ""
-    ipv6_flag = "P"
-    if data["sysctlid"].find("ipv6") >= 0:
-        ipv6_flag = "I"
-    data["flags"] = "SR" + ipv6_flag
-    return data
-
-
-@template(["anaconda", "ansible", "bash", "oval", "puppet"])
-def package_removed(data, lang):
-    return data
-
-
-@template(["ansible", "bash", "oval"])
-def sebool(data, lang):
-    sebool_bool = data.get("sebool_bool", None)
-    if sebool_bool is not None and sebool_bool not in ["true", "false"]:
-        raise ValueError(
-            "ERROR: key sebool_bool in rule {0} contains forbidden "
-            "value '{1}'.".format(data["_rule_id"], sebool_bool)
+class Template():
+    def __init__(self, template_root_directory, name):
+        self.template_root_directory = template_root_directory
+        self.name = name
+        self.template_path = os.path.join(self.template_root_directory, self.name)
+        self.preprocessing_file_path = os.path.join(self.template_path, preprocessing_file_name)
+        self.langs = []
+        for lang in languages:
+            langfilename = lang + ".template"
+            if os.path.exists(os.path.join(self.template_path, langfilename)):
+                self.langs.append(lang)
+
+    def preprocess(self, parameters, lang):
+        preprocess_mod = None  # temporarily imported module containing preprocessing function
+        spec = importlib.util.spec_from_file_location(
```

Comment:
I have an idea about the `templates.py` file in the template directory. When I don't want to `preprocess` the data aka I don't need to change the input in any way, why not to just skip this logic (e.g capturing some exception that the file doesn't exist) and input the raw data to the jinja template? This way we don't need to have dummy `templates.py` in every template directory.

Reply:
I am not against this idea. But can't we use these files also for documentation purposes? I am talking about your readthedocs initiative.

Reply:
I'm not sure if it adds real value having an empty function without any comments

Reply:
I like the idea of being able to skip e.g. the Python file, and I think that concerning documentation, we could aim at having e.g. a `template.rst` file that would contain the relevant snippet. In other words, I think that we don't need a function as a carrier of documentation.